File: backend/src/services/web.service.ts
import { getFirecrawl } from "../scraper.js";
import { WebsiteModel, SitemapModel } from "../models.js";
import type { VectorSearchHit } from "../types.js";
import { getEmbedding } from "./ai.service.js";

type CrawlLink = {
  url: string;
  title: string;
};

const MAP_URL_LIMIT = Number(process.env.FIRECRAWL_MAP_PAGE_LIMIT || 300);
const INDEX_WORKER_COUNT = 2;

const normalizeUrl = (value: string) => {
  try {
    return new URL(value).toString();
  } catch {
    try {
      return new URL(`https://${value}`).toString();
    } catch {
      return null;
    }
  }
};

const isSameSiteUrl = (value: string, domain: string) => {
  try {
    const parsed = new URL(value);
    return parsed.hostname === domain || parsed.hostname.endsWith(`.${domain}`);
  } catch {
    return false;
  }
};

const runWorkerPool = async <T>(
  items: T[],
  workerCount: number,
  worker: (item: T, workerIndex: number, itemIndex: number) => Promise<void>,
) => {
  let nextIndex = 0;
  const workers = Array.from({ length: workerCount }, (_, workerIndex) =>
    (async () => {
      while (true) {
        const itemIndex = nextIndex;
        nextIndex += 1;
        if (itemIndex >= items.length) {
          break;
        }

        await worker(items[itemIndex]!, workerIndex, itemIndex);
      }
    })(),
  );

  await Promise.all(workers);
};

const dedupeLinks = (links: CrawlLink[]) => {
  return Array.from(new Map(links.map((link) => [link.url, link])).values());
};

const tokenizeForRanking = (value: string) =>
  value
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, " ")
    .split(/\s+/)
    .filter((token) => token.length > 2);

const scoreLinkForQuestion = (question: string, link: CrawlLink) => {
  const questionTokens = tokenizeForRanking(question);
  if (questionTokens.length === 0) return 0;

  const combined = `${link.title} ${link.url}`.toLowerCase();
  let score = 0;

  for (const token of questionTokens) {
    if (combined.includes(token)) {
      score += token.length >= 5 ? 3 : 1;
    }
  }

  return score;
};

export const rankLinksForQuestion = (
  question: string,
  links: CrawlLink[],
  limit = 10,
) => {
  return [...links]
    .map((link) => ({ ...link, score: scoreLinkForQuestion(question, link) }))
    .sort((a, b) => b.score - a.score)
    .slice(0, limit)
    .map(({ score: _score, ...link }) => link);
};

export const extractUrlsFromText = (text: string) => {
  const matches = text.match(/https?:\/\/[^\s"'<>]+/gi) ?? [];
  const normalized = matches
    .map((url) => url.replace(/[),.;]+$/g, ""))
    .map((url) => normalizeUrl(url))
    .filter((url): url is string => Boolean(url));

  return Array.from(new Set(normalized));
};

export const indexUrls = async (
  domain: string,
  links: CrawlLink[],
  options?: { workerCount?: number; maxUrls?: number },
) => {
  const uniqueLinks = dedupeLinks(
    links
      .map((link) => ({
        url: normalizeUrl(link.url) ?? link.url,
        title: link.title || "",
      }))
      .filter((link) => Boolean(link.url) && isSameSiteUrl(link.url, domain)),
  ).slice(0, options?.maxUrls ?? MAP_URL_LIMIT);

  if (uniqueLinks.length === 0) {
    return { total: 0, success: 0, failed: 0 };
  }

  const workerCount = Math.max(
    1,
    Math.min(options?.workerCount ?? INDEX_WORKER_COUNT, uniqueLinks.length),
  );

  let success = 0;
  let failed = 0;

  await runWorkerPool(uniqueLinks, workerCount, async (link, workerIndex) => {
    try {
      const textToEmbed = link.title || link.url.split("/").pop() || domain;
      const vector = await getEmbedding(textToEmbed, workerIndex);

      await SitemapModel.findOneAndUpdate(
        { url: link.url },
        { domain, url: link.url, title: link.title || "", embedding: vector },
        { upsert: true },
      );

      success += 1;
    } catch (error) {
      failed += 1;
      const errorMessage =
        error instanceof Error ? error.message : String(error);
      console.error(
        `[MAP][INDEX][ERROR] domain=${domain} url=${link.url} error=${errorMessage}`,
      );
    }
  });

  return { total: uniqueLinks.length, success, failed };
};

const cosineSimilarity = (left: number[], right: number[]) => {
  if (left.length === 0 || right.length === 0) return 0;

  const size = Math.min(left.length, right.length);
  let dot = 0;
  let leftMagnitude = 0;
  let rightMagnitude = 0;

  for (let index = 0; index < size; index += 1) {
    const leftValue = left[index] ?? 0;
    const rightValue = right[index] ?? 0;
    dot += leftValue * rightValue;
    leftMagnitude += leftValue * leftValue;
    rightMagnitude += rightValue * rightValue;
  }

  if (leftMagnitude === 0 || rightMagnitude === 0) {
    return 0;
  }

  return dot / (Math.sqrt(leftMagnitude) * Math.sqrt(rightMagnitude));
};

export const vectorSearchWithinUrls = async (
  questionVector: number[],
  domain: string,
  urls: string[],
  limit = 6,
): Promise<VectorSearchHit[]> => {
  const uniqueUrls = Array.from(
    new Set(
      urls
        .map((url) => normalizeUrl(url) ?? url)
        .filter(
          (url): url is string => Boolean(url) && isSameSiteUrl(url, domain),
        ),
    ),
  );

  if (uniqueUrls.length === 0) {
    return [];
  }

  const docs = (await SitemapModel.find({
    domain,
    url: { $in: uniqueUrls },
  }).lean()) as Array<{ url: string; embedding: number[] }>;

  return docs
    .map((doc) => ({
      url: doc.url,
      score: cosineSimilarity(questionVector, doc.embedding ?? []),
    }))
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
};

const isEmbeddingQuotaError = (error: unknown) => {
  const message =
    error instanceof Error
      ? error.message
      : String((error as { message?: unknown })?.message ?? error ?? "");

  return /quota exceeded|too many requests|429|embed_content_free_tier_requests|gemini-embedding/i.test(
    message,
  );
};

export const mapNewWebsite = async (domain: string, websiteUrl: string) => {
  const startedAt = Date.now();
  console.log(`[MAP][START] domain=${domain} baseUrl=${websiteUrl}`);

  const mapAttempt = async (preference: "primary" | "secondary") => {
    const firecrawl = getFirecrawl({ preference });
    return firecrawl.map(websiteUrl, {
      sitemap: "include",
      ignoreQueryParameters: true,
      includeSubdomains: false,
      limit: MAP_URL_LIMIT,
      timeout: 120000,
    });
  };

  let mapRes;
  try {
    mapRes = await mapAttempt("primary");
  } catch (error) {
    console.warn(
      `[MAP][CRAWL][RETRY] domain=${domain} mode=secondary reason=${error instanceof Error ? error.message : String(error)}`,
    );
    mapRes = await mapAttempt("secondary");
  }

  const rawCount = mapRes.links?.length ?? 0;
  console.log(`[MAP][CRAWL] domain=${domain} discoveredLinks=${rawCount}`);

  if (!mapRes.links || mapRes.links.length === 0) {
    const elapsedMs = Date.now() - startedAt;
    console.warn(
      `[MAP][DONE] domain=${domain} status=no-links total=0 success=0 failed=0 durationMs=${elapsedMs}`,
    );
    return;
  }

  const normalizedLinks: CrawlLink[] = mapRes.links
    .map((item: any) => {
      if (typeof item === "string") {
        return { url: item, title: "" };
      }

      if (item && typeof item.url === "string") {
        return {
          url: item.url,
          title: typeof item.title === "string" ? item.title : "",
        };
      }

      return null;
    })
    .filter((item): item is CrawlLink => Boolean(item?.url));

  const uniqueLinks = dedupeLinks(normalizedLinks).slice(0, MAP_URL_LIMIT);
  console.log(
    `[MAP][CRAWL] domain=${domain} uniqueLinks=${uniqueLinks.length}`,
  );

  const indexResult = await indexUrls(domain, uniqueLinks, {
    workerCount: INDEX_WORKER_COUNT,
    maxUrls: MAP_URL_LIMIT,
  });

  await WebsiteModel.updateOne(
    { domain },
    { isMapped: true },
    { upsert: true },
  );

  const elapsedMs = Date.now() - startedAt;
  console.log(
    `[MAP][DONE] domain=${domain} total=${indexResult.total} success=${indexResult.success} failed=${indexResult.failed} durationMs=${elapsedMs}`,
  );
};

export const vectorSearch = async (
  questionVector: number[],
  domain: string,
): Promise<VectorSearchHit[]> => {
  return await SitemapModel.aggregate<VectorSearchHit>([
    {
      $vectorSearch: {
        index: "vector_index",
        path: "embedding",
        queryVector: questionVector,
        filter: { domain },
        numCandidates: 50,
        limit: 1,
      },
    },
    { $project: { url: 1, domain: 1, score: { $meta: "vectorSearchScore" } } },
    { $match: { domain } },
  ]);
};

File: backend-java/src/main/java/com/universal/webai/services/WebService.java
package com.universal.webai.services;

import com.universal.webai.models.Sitemap;
import com.universal.webai.models.Website;
import com.universal.webai.repositories.SitemapRepository;
import com.universal.webai.repositories.WebsiteRepository;
import com.universal.webai.scraper.Scraper;
import com.universal.webai.types.VectorSearchHit;
import org.bson.Document;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.aggregation.Aggregation;
import org.springframework.data.mongodb.core.aggregation.AggregationOperation;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Web / sitemap service — crawls websites, stores embeddings, and performs vector search.
 *
 * <p>Equivalent to {@code web.service.ts} in the Node.js backend.
 * Provides:
 * <ul>
 *   <li>{@link #mapNewWebsite} — async background full-site crawl + embedding storage</li>
 *   <li>{@link #vectorSearch} — MongoDB Atlas {@code $vectorSearch} aggregation</li>
 * </ul>
 */
@Service
public class WebService {

    private static final Logger log = LoggerFactory.getLogger(WebService.class);

    private final Scraper scraper;
    private final AiService aiService;
    private final WebsiteRepository websiteRepository;
    private final SitemapRepository sitemapRepository;
    private final MongoTemplate mongoTemplate;

    public WebService(
            Scraper scraper,
            AiService aiService,
            WebsiteRepository websiteRepository,
            SitemapRepository sitemapRepository,
            MongoTemplate mongoTemplate) {
        this.scraper = scraper;
        this.aiService = aiService;
        this.websiteRepository = websiteRepository;
        this.sitemapRepository = sitemapRepository;
        this.mongoTemplate = mongoTemplate;
    }

    /**
     * Crawls all pages of a website, embeds each URL, and persists the results.
     *
     * <p>Equivalent to {@code mapNewWebsite(domain, websiteUrl)} in {@code web.service.ts}.
     * This method is annotated {@link Async} so it runs in a background thread and does not
     * block the initial query response — matching the {@code void mapNewWebsite(...)} fire-and-forget
     * pattern in the Node.js code.
     *
     * @param domain     hostname of the site (e.g. {@code "example.com"})
     * @param websiteUrl root URL of the site to crawl
     */
    @Async
    public void mapNewWebsite(String domain, String websiteUrl) {
        long startedAt = System.currentTimeMillis();
        log.info("[MAP][START] domain={} baseUrl={}", domain, websiteUrl);

        // Discover all page URLs via the Firecrawl map endpoint
        List<String> rawLinks = scraper.mapWebsite(websiteUrl);
        int rawCount = rawLinks.size();
        log.info("[MAP][CRAWL] domain={} discoveredLinks={}", domain, rawCount);

        if (rawLinks.isEmpty()) {
            log.warn("[MAP][DONE] domain={} status=no-links total=0 success=0 failed=0 durationMs={}",
                    domain, System.currentTimeMillis() - startedAt);
            return;
        }

        // De-duplicate links (preserving insertion order)
        List<String> uniqueLinks = rawLinks.stream()
                .distinct()
                .collect(Collectors.toList());
        log.info("[MAP][CRAWL] domain={} uniqueLinks={}", domain, uniqueLinks.size());

        int total = uniqueLinks.size();
        int[] processed = {0}, success = {0}, failed = {0};
        Map<String, Integer> errorBuckets = new LinkedHashMap<>();

        // Embed each URL and upsert the sitemap document
        for (int index = 0; index < total; index++) {
            String url = uniqueLinks.get(index);
            int displayIndex = index + 1;

            if (url == null || url.isBlank()) {
                failed[0]++;
                processed[0]++;
                log.warn("[MAP][SKIP] domain={} index={}/{} reason=missing-url progress={}/{}",
                        domain, displayIndex, total, processed[0], total);
                continue;
            }

            try {
                // Build text to embed: use URL slug as a proxy for page title
                // (same heuristic as the Node.js backend when no title is available)
                String[] segments = url.split("/");
                String slug = segments.length > 0 ? segments[segments.length - 1] : domain;
                String textToEmbed = slug.isBlank() ? domain : slug;

                List<Double> vector = aiService.getEmbedding(textToEmbed);

                // Upsert the sitemap entry (equivalent to findOneAndUpdate with upsert:true)
                upsertSitemap(domain, url, textToEmbed, vector);

                success[0]++;
                processed[0]++;
                log.info("[MAP][OK] domain={} index={}/{} progress={}/{} url={}",
                        domain, displayIndex, total, processed[0], total, url);

            } catch (Exception err) {
                failed[0]++;
                processed[0]++;
                String errorMessage = err.getMessage() != null ? err.getMessage() : err.toString();
                String errorKey = errorMessage.substring(0, Math.min(errorMessage.length(), 120));
                errorBuckets.merge(errorKey, 1, Integer::sum);
                log.error("[MAP][ERROR] domain={} index={}/{} progress={}/{} url={} error={}",
                        domain, displayIndex, total, processed[0], total, url, errorMessage);
            }
        }

        // Mark the domain as fully mapped so subsequent requests skip the crawl
        upsertWebsite(domain, true);

        long elapsedMs = System.currentTimeMillis() - startedAt;
        log.info("[MAP][DONE] domain={} total={} success={} failed={} durationMs={}",
                domain, total, success[0], failed[0], elapsedMs);

        if (failed[0] > 0) {
            String errorSummary = errorBuckets.entrySet().stream()
                    .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                    .limit(10)
                    .map(e -> e.getValue() + "x " + e.getKey())
                    .collect(Collectors.joining(" | "));
            log.warn("[MAP][ERROR_SUMMARY] domain={} topErrors={}", domain, errorSummary);
        }
    }

    /**
     * Performs a vector similarity search against the {@code sitemaps} collection.
     *
     * <p>Equivalent to {@code vectorSearch(questionVector, domain)} in {@code web.service.ts}.
     * Uses the MongoDB Atlas {@code $vectorSearch} aggregation stage (requires a
     * {@code vector_index} index on the {@code embedding} field).
     *
     * @param questionVector dense embedding of the user's question
     * @param domain         hostname to restrict the search to
     * @return list of matching {@link VectorSearchHit} records
     */
    public List<VectorSearchHit> vectorSearch(List<Double> questionVector, String domain) {
        // Build the $vectorSearch stage as a raw BSON document
        // (Spring Data MongoDB does not yet provide a typed builder for this stage)
        Document vectorSearchStage = new Document("$vectorSearch", new Document()
                .append("index", "vector_index")
                .append("path", "embedding")
                .append("queryVector", questionVector)
                .append("numCandidates", 50)
                .append("limit", 1));

        // $project: return url, domain and the similarity score
        Document projectStage = new Document("$project", new Document()
                .append("url", 1)
                .append("domain", 1)
                .append("score", new Document("$meta", "vectorSearchScore")));

        // $match: restrict to the queried domain
        Document matchStage = new Document("$match", new Document("domain", domain));

        List<Document> pipeline = List.of(vectorSearchStage, projectStage, matchStage);

        List<Document> rawResults = mongoTemplate.getDb()
                .getCollection("sitemaps")
                .aggregate(pipeline)
                .into(new ArrayList<>());

        return rawResults.stream()
                .map(doc -> new VectorSearchHit(
                        doc.getString("url"),
                        doc.getDouble("score") != null ? doc.getDouble("score") : 0.0))
                .collect(Collectors.toList());
    }

    // ── Private helpers ────────────────────────────────────────────────────────

    /**
     * Upserts a website record (equivalent to
     * {@code WebsiteModel.updateOne({ domain }, { domain, isMapped }, { upsert: true })}).
     */
    private void upsertWebsite(String domain, boolean isMapped) {
        mongoTemplate.upsert(
                Query.query(Criteria.where("domain").is(domain)),
                Update.update("domain", domain).set("isMapped", isMapped),
                Website.class);
    }

    /**
     * Upserts a sitemap (page) record (equivalent to
     * {@code SitemapModel.findOneAndUpdate({ url }, { domain, url, title, embedding }, { upsert: true })}).
     */
    private void upsertSitemap(String domain, String url, String title, List<Double> embedding) {
        mongoTemplate.upsert(
                Query.query(Criteria.where("url").is(url)),
                Update.update("url", url)
                        .set("domain", domain)
                        .set("title", title)
                        .set("embedding", embedding),
                Sitemap.class);
    }
}
